.github/actions/setup-environment/action.yml

```diff
@@ -53,6 +55,22 @@ runs:
         ./eden config set default --key=eve.cpu --value=2
       shell: bash
       working-directory: "./eden"
+    - name: Restore cache if specified
+      if: |
+        inputs.eve_cache != "" &&
+        inputs.eve_image != ""
+      uses: action/cache/restore@v3
+      with:
+        path: inputs.eve_image
+        key: inputs.eve_cache
```

Comment:
I don't understand this. I had thought that we were passing in a file path explicitly. Are we instead expecting the tar file to exist as an object saved in the GHA cache?

Maybe a write-up on how to use this with a sample, e.g. in a README?

Reply:
I had thought that in order to pass objects between jobs you need to either use cache or artifacts. Artifacts are published and stored for 90 days, cache is not published and GitHub deletes previous entries if total amount of cache is more than 10Gbs. Artifacts also rebuild each time. 

In our case, before [test_suite_pr](https://github.com/lf-edge/eve/blob/master/.github/workflows/eden.yml#L48)  job we need to store EVE danger image somewhere in cache and then restore this cache in our reusable workflow. Unfortunately, you can't add additional steps before invoking reusable workflow, otherwise we could have just do `docker load` before invoking tests workflow. 

In order to restore cache from test workflow, one needs key to restore this cache from and what files exactly we want to restore. We, of course can use [wildcard](https://github.com/actions/toolkit/tree/main/packages/glob) to restore all the files from cache, but we will need to access this file afterwards.

I though it would make sense to have a convention of naming our archive same as we name our eve image (would make sense to add tar.gz extension though) 

Hope I answered your question @deitch, I will add this paragraph in GitHub actions [doc](https://github.com/lf-edge/eden/blob/master/docs/github-actions.md)

Reply:
> I had thought that in order to pass objects between jobs you need to either use cache or artifacts. Artifacts are published and stored for 90 days, cache is not published and GitHub deletes previous entries if total amount of cache is more than 10Gbs. Artifacts also rebuild each time

Between jobs, yes; but not between steps.

I would keep the cache out of this. Have it entirely be a file that it uploads. Why have this action worry about cache or artifact or whatever. Let that be an independent step. Something like this:

```yaml
  test_suite_pr:
    runs-on: ubuntu-latest
    steps:
      - name: retrieve eve image
        id: eve_image_cache_restore
        uses: actions/cache/restore@v3
        with:
          path: path/to/eve.tar
          key: eve-tar-amd64-${{ github.sha }}
          fail-on-cache-miss: true
    - if: github.event.review.state == 'approved'
      uses: lf-edge/eden/.github/workflows/test.yml@0.9.3
      with:
        eve_image: "evebuild/danger:pr${{ github.event.pull_request.number  }}"
        eve_image_at: path/to/eve.tar
```

Reply:
I kind of messed up the indenting, but I think it still shows it.

Reply:
I though that you can't do any steps before invoking reusable workflow @deitch. If we can, we could actually do docker load before invoking test.yml workflow on EVE side and then you don't have to change anything on Eden side (since job is executed within one machine eden will use local image)

Reply:
I.e.
```
  test_suite_pr:
    runs-on: ubuntu-latest
    if: github.event.review.state == 'approved'
    steps:
      - name: retrieve eve image
        id: eve_image_cache_restore
        uses: actions/cache/restore@v3
        with:
          path: path/to/eve.tar
          key: eve-tar-amd64-${{ github.sha }}
          fail-on-cache-miss: true
      - name: load eve image to docker
         run: |
            docker load -q -i path/to/eve.tar
      - uses: lf-edge/eden/.github/workflows/test.yml@0.9.3
         with:
           eve_image: "evebuild/danger:pr${{ github.event.pull_request.number  }}"
```

Assuming that image we loaded to docker has name `eve_image`

Reply:
Yes, of course. That would be a prerequisite.

Reply:
> I though that you can't do any steps before invoking reusable workflow @deitch. If we can, we could actually do docker load before invoking test.yml workflow on EVE side and then you don't have to change anything on Eden side (since job is executed within one machine eden will use local image)

I had thought you could, but now I am questioning it as well. Let's double-check that.

Reply:
it should be ok. We are not using [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) here, but [composite actions](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action).

Reply:
On EVE side [here](https://github.com/lf-edge/eve/blob/09aa0fe1026e263763d390130ee40c58a878b152/.github/workflows/eden.yml#L51) we use reusable [workflow](https://github.com/lf-edge/eden/blob/ed507793968a2005212d589d6c3d88824783a9a7/.github/workflows/test.yml#L9), that's why this PR. Making eden tests as composite actions will make us loose visibility on what test succeeded and what tests failed.